### PR TITLE
refactor face detection and recognition

### PR DESCRIPTION
**Pull request by UnaNancyOwen**
This pull request is proposal implementation of high level api for face detection and recognition.
Please see https://github.com/opencv/opencv/issues/20874 for this proposal deteal.

- [x] Re-implementation face detection (`cv::dnn::FaceDetectionModel_YN`) and recognition (`cv::dnn::FaceRecognitionModel_SF`) to high level api extended from `cv::dnn::Model`.
- [x] Add support for [OpenCV Face Detector](https://github.com/opencv/opencv/tree/5.x/samples/dnn/face_detector) to face detection (`cv::dnn::FaceDetectionModel_SSD`).
- [x] Change the output format to be user-friendly from `cv::FaceDetectorYN` style.
       (Separate results to bounding box, confidence, and landmark.)
- [x] Prepare sample programs and tutorials. 
- [x] Add test code.
- [x] Remove `cv::FaceDetectorYN` and `cv::FaceRecognizerSF` from objdetect module.
       (Note: I think the CI for iOS will pass once this task is complete.)
- [x] Update branch after merge https://github.com/opencv/opencv/pull/21607 and update 5.x branch
- [x] Fix model file name (yunet-202202.onnx)

Users can use face detection and recognition in the following samples with this proposal.  
It is as easy to use with unified usability as the other high level apis.

* [cv::dnn::FaceDetectionModel_SSD](https://gist.github.com/UnaNancyOwen/f887d834a0d8d68e59f0de25ac40240c#file-face_detection_ssd-cpp)/[cv.dnn_FaceDetectionModel_SSD](https://gist.github.com/UnaNancyOwen/f887d834a0d8d68e59f0de25ac40240c#file-face_detection_ssd-py)
* [cv::dnn::FaceDetectionModel_YN](https://gist.github.com/UnaNancyOwen/3b4b9ce7a73c709c05e2fd949080d35e#file-face_detector-cpp)/[cv.dnn_FaceDetectionModel_YN](https://gist.github.com/UnaNancyOwen/3b4b9ce7a73c709c05e2fd949080d35e#file-face_detector-py)
* [cv::dnn::FaceRecognitionModel_SF](https://gist.github.com/UnaNancyOwen/8817c63c7a1acb7817812cae561765bc#file-face_recognizer-cpp)/[cv.dnn_FaceRecognitionModel_SF](https://gist.github.com/UnaNancyOwen/8817c63c7a1acb7817812cae561765bc#file-face_recognizer-py)

What do you think this proposal? Thanks,

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by UnaNancyOwen** [comment]
This error occurs in OpenCV 5.x. This error not occurs in OpenCV 4.x. 
However, the inference will be executed correctly and face is detected correctly in each versions.
This error would be a separate problem from this Pull Request.

```
[ERROR:0@109.908] global /build/precommit_linux64/5.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (2516) parseShape DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 243 [ 0 0 0 51 ]
[ERROR:0@109.908] global /build/precommit_linux64/5.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (2516) parseShape DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 250 [ 0 0 0 34 ]
[ERROR:0@109.908] global /build/precommit_linux64/5.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (2516) parseShape DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 257 [ 0 0 0 34 ]
[ERROR:0@109.908] global /build/precommit_linux64/5.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (2516) parseShape DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 264 [ 0 0 0 51 ]
[ERROR:0@109.909] global /build/precommit_linux64/5.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (2516) parseShape DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 297 [ 0 -4 ]
```

**Comment by UnaNancyOwen** [comment]
> This error occurs in OpenCV 5.x, but not occurs in OpenCV 4.x. However, the inference will be executed correctly and face is detected correctly in both versions. This error would be a separate problem from this pull request.
> 
> OpenCV 4.x	OpenCV 5.x
> cv::FaceDetectionModel_YN
> (new from this pull request)	-	occurs
> cv::FaceDetectorYN
> (old)	not occurs	occurs
> ```
> [ERROR:0@109.908] global /build/precommit_linux64/5.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (2516) parseShape DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 243 [ 0 0 0 51 ]
> [ERROR:0@109.908] global /build/precommit_linux64/5.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (2516) parseShape DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 250 [ 0 0 0 34 ]
> [ERROR:0@109.908] global /build/precommit_linux64/5.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (2516) parseShape DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 257 [ 0 0 0 34 ]
> [ERROR:0@109.908] global /build/precommit_linux64/5.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (2516) parseShape DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 264 [ 0 0 0 51 ]
> [ERROR:0@109.909] global /build/precommit_linux64/5.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (2516) parseShape DNN/ONNX(Shape): dynamic 'zero' shapes are not supported, input 297 [ 0 -4 ]
> ```

It is fixed by https://github.com/opencv/opencv/pull/21607 .

**Comment by alalek** [comment]
> but not occurs in OpenCV 4.x.

This is not correct. There is the same problem in 4.x.
They are just messages without triggered error exceptions (disabled).

**Comment by UnaNancyOwen** [comment]
> This is not correct. There is the same problem in 4.x.
> They are just messages without triggered error exceptions (disabled).

Thanks, I understood it.

**Comment by UnaNancyOwen** [comment]
@alalek This pull request are ready to review. Thanks,

File: modules/dnn/src/model.cpp [hunk 1]
```diff
@@ -1499,5 +1499,770 @@ int TextDetectionModel_DB::getMaxCandidates() const
...
+            {
+                const int x = static_cast<int>(output.at<float>(j));
+                const int y = static_cast<int>(output.at<float>(j + 1));
+                landmark.push_back(Point(x, y));
+            }
+            landmarks.push_back(landmark);
+
+            const float conficence = output.at<float>(14);
+            confidences.push_back(conficence);
+        }
```
Comment (fengyuentau) [inline modules/dnn/src/model.cpp:1758]: This block of formatting outputs can be blended into `postProcess` to reduce redundancy.

File: modules/dnn/test/test_model.cpp [hunk 2]
```diff
@@ -738,6 +914,178 @@ TEST_P(Test_Model, TextDetectionByEAST)
...
+    );
+}
+
+TEST_P(Test_Model, FaceDetectionByYN)
+{
+    if (target == DNN_TARGET_OPENCL_FP16)
+        applyTestTag(CV_TEST_TAG_DNN_SKIP_OPENCL_FP16);
+
+    // Weight
+    std::string face_detection_weight_path = _tf("onnx/models/yunet-202109.onnx", false);
```
Comment (fengyuentau) [inline modules/dnn/test/test_model.cpp:967]: [A fix](https://github.com/opencv/opencv/pull/21607) is on-they-way. A new version of YuNet should be used instead after the fix is merged.

File: modules/dnn/test/test_model.cpp [hunk 3]
```diff
@@ -738,6 +914,178 @@ TEST_P(Test_Model, TextDetectionByEAST)
...
+    );
+}
+
+TEST_P(Test_Model, FaceRecognitionBySF)
+{
+    if (target == DNN_TARGET_OPENCL_FP16)
+        applyTestTag(CV_TEST_TAG_DNN_SKIP_OPENCL_FP16);
+
+    // Weight
+    std::string face_detection_weight_path = _tf("onnx/models/yunet-202109.onnx", false);
```
Comment (fengyuentau) [inline modules/dnn/test/test_model.cpp:1054]: Same as above.

**Review by fengyuentau — COMMENTED** [review]
Thanks for the contribution! Please have a look at the comments.

File: modules/dnn/test/test_model.cpp (hunk 3 above)
Comment (sturkmen72) [reply modules/dnn/test/test_model.cpp:1054]: new updates on [tutorial page](https://docs.opencv.org/4.x/d0/dd4/tutorial_dnn_face.html) have been ignored

**Comment by alalek** [comment]
BTW, 5.x branch doesn't have updates from 4.x since December 2021

File: modules/dnn/src/model.cpp (hunk 1 above)
Comment (UnaNancyOwen) [reply modules/dnn/src/model.cpp:1758]: I ported it as-is from your original implement. However as you say, it is more efficient to incorporate it in post process function. I'll consider and implement it. Thanks,

File: modules/dnn/test/test_model.cpp (hunk 2 above)
Comment (UnaNancyOwen) [reply modules/dnn/test/test_model.cpp]: Okay, I'll fix to load new model after that pull request merged.

File: modules/dnn/src/model.cpp (hunk 1 above)
Comment (UnaNancyOwen) [reply modules/dnn/src/model.cpp]: I've implemented it! 👍

**Comment by UnaNancyOwen** [comment]
> BTW, 5.x branch doesn't have updates from 4.x since December 2021

@alalek Thanks, I will operate to update the branch of this pull request when the 5.x branch is updated. However, please marge this pull requests https://github.com/opencv/opencv/pull/21607 (it is target to 4.x) before update 5.x branch.

File: modules/dnn/test/test_model.cpp [hunk 4]
```diff
@@ -738,6 +914,178 @@ TEST_P(Test_Model, TextDetectionByEAST)
...
+                iss_x >> x;
+                getline(ss, part, ' ');
+                std::istringstream iss_y(part);
+                iss_y >> y;
+                points.push_back(Point(x, y));
+            }
+            gt_landmarks.push_back(points);
+        }
+
+        // There are 7 faces in the addams-family.png, but yunet-202109.onnx can't detect the last one.
```
Comment (sturkmen72) [inline modules/dnn/test/test_model.cpp:1026]: just a note: all faces found if `score_threshold = 0.85`

Comment (UnaNancyOwen) [reply modules/dnn/test/test_model.cpp]: Oh, It is my mistake. The threshold was not applied correctly. I will fix it. Thanks,

Comment (UnaNancyOwen) [reply modules/dnn/test/test_model.cpp]: fixed it!

**Comment by UnaNancyOwen** [comment]
@fengyuentau yunet-202202.onnx requere to fix input size to 120x160?

**Comment by fengyuentau** [comment]
> @fengyuentau yunet-202202.onnx requere to fix input size to 120x160?

Yes for ONNX and no for OpenCV DNN. OpenCV DNN does not support ONNX models of dynamic input shape, but it runs the graph on the actual input shape.

**Comment by UnaNancyOwen** [comment]
@fengyuentau Thanks, I have updated the model, but the test fails. If you notice anything, please give a comment.

**Comment by UnaNancyOwen** [comment]
@fengyuentau In the 5.x/HEAD, `cv::FaceDetectorYN` (objdetect api) with yunet_120x160.onnx does not work correctly regardless of this pull request? Please confirm.

**Comment by fengyuentau** [comment]
I tested the latest 5.x and the unit tests of `cv::FaceDetectorYN` all passed. BTW, latest 5.x is using yunet_120x160: 
https://github.com/opencv/opencv/blob/aa5055261f09b11bca97c39b644085a5ef5f4af6/modules/objdetect/test/test_face.cpp#L81

https://github.com/opencv/opencv/blob/aa5055261f09b11bca97c39b644085a5ef5f4af6/modules/objdetect/test/test_face.cpp#L181

**Comment by UnaNancyOwen** [comment]
@fengyuentau Sorry, It was my mistake and misunderstanding. I have solved this problem. Thanks,

**Comment by UnaNancyOwen** [comment]
@alalek @fengyuentau All tests are now green. Please review it.

**Pull request closed without merging**